contracts/ibc-callbacks/src/contract.rs

```diff
@@ -0,0 +1,174 @@
+    }
+}
+
+/// This is the entrypoint that is called by the source chain when a callbacks-enabled IBC message
+/// is acknowledged or times out.
+// #[entry_point]
+pub fn ibc_source_chain_callback(
+    deps: DepsMut,
+    _env: Env,
+    msg: IbcSourceChainCallback,
+) -> StdResult<IbcBasicResponse> {
+    let mut counts = load_stats(deps.storage)?;
+
+    match msg {
+        IbcSourceChainCallback::Acknowledgement(ack) => {
+            // increment the counter
+            counts.ibc_ack_callbacks.push(ack);
+        }
+        IbcSourceChainCallback::Timeout(timeout) => {
+            // increment the counter
+            counts.ibc_timeout_callbacks.push(timeout);
+        }
+    }
+
+    save_stats(deps.storage, &counts)?;
+
+    Ok(IbcBasicResponse::new().add_attribute("action", "ibc_source_chain_callback"))
+}
+
+// TODO: move all the types below to cosmwasm-std when the everything is ready
+
+/// This is just a type representing the data that has to be sent with the IBC message to receive
+/// callbacks. It should be serialized to JSON and sent with the IBC message.
+/// The specific field to send it in can vary depending on the IBC message,
+/// but is usually the `memo` field by convention.
+///
+/// See [`IbcSourceChainCallback`] for more details.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct IbcCallbackData {
+    // using private fields to force use of the constructors
+    src_callback: Option<IbcSrcCallback>,
+    dst_callback: Option<IbcDstCallback>,
+}
+
+impl IbcCallbackData {
+    /// Use this if you want to execute callbacks on both the source and destination chain.
+    ///
+    /// In the first prototype, we only support receiving callbacks on the source chain.
+    pub fn both(src_callback: IbcSrcCallback, dst_callback: IbcDstCallback) -> Self {
+        IbcCallbackData {
+            src_callback: Some(src_callback),
+            dst_callback: Some(dst_callback),
+        }
+    }
+
+    /// Use this if you want to execute callbacks on the source chain, but not the destination chain.
+    pub fn source(src_callback: IbcSrcCallback) -> Self {
+        IbcCallbackData {
+            src_callback: Some(src_callback),
+            dst_callback: None,
+        }
+    }
+
+    /// Use this if you want to execute callbacks on the destination chain, but not the source chain.
+    ///
+    /// In the first prototype, we only support receiving callbacks on the source chain.
+    pub fn destination(dst_callback: IbcDstCallback) -> Self {
+        IbcCallbackData {
+            src_callback: None,
+            dst_callback: Some(dst_callback),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct IbcSrcCallback {
+    /// The source chain address that should receive the callback.
+    /// You probably want to put `env.contract.address` here.
+    pub address: Addr,
```

Comment:
Oh but you can inform a different contract instead of yourself?

Reply:
The spec allows that, but it would get caught by the check in wasmd mentioned above.
If we want to support calling a different contract, we should not do that check and instead forward the sender address to the contract and let it handle the check.